### ISSUE TITLE
Support testing multidimensional np.random.poisson sampling

### ISSUE THREAD

**Issue by daria-shaw** [numba/numba#9853]
## Feature request

The [overloaded poisson implementations](https://github.com/numba/numba/blob/8348b5d844d3f67ce5dc41d68a09a10dd5b6b927/numba/cpython/randomimpl.py#L1633-L1722) support the following types for size:
* None
* ()
* (int,)

I'd like to add a new overload poisson_impl3 which supports (int,int) or an arbitrary size of tuple, enabling multidimensional poisson sampling.

This is currently supported by the [numpy implementation](https://numpy.org/doc/2.0/reference/random/generated/numpy.random.poisson.html):
```
size
int or tuple of ints, optional

Output shape. 
If the given shape is, e.g., (m, n, k), then m * n * k samples are drawn. 
If size is None (default), a single value is returned if lam is a scalar. 
Otherwise, np.array(lam).size samples are drawn.

```

**Comment by svenrahmann** [comment]
This would be nice, but can actually be easily worked around by specifying a single integer as m * n * k (instead of the tuple (m, n, k) and then using `.reshape(m, n, k)` on the result, which is a cheap operation.

**Comment by gmarkall** [comment]
@daara-s Many thanks for the feature request - were you suggesting you'd like to make a PR for this? If so, feel free to, or let us know if you'd like some further guidance.

@svenrahmann Thanks for suggesting a workaround for the meantime!

**Comment by daara-s** [comment]
@gmarkall yeah I’d like to give it a go with a PR. Will try to get an initial draft out this weekend.

**Comment by esc** [comment]
Following up on this. I have started to review #9875 but ran into some concerns. I came to the conclusion that the feature multidimensional sampling of `np.random.poisson` is/was already possible. I just tried commit:

62d1c4ad474a107bcd24f78ac85a9bb8b2c7c8c7

And I got

```
In [1]: from numpy.random import poisson

In [2]: from numba import njit

In [3]: @njit
   ...: def subroutine(lam, size):
   ...:     return poisson(lam, size)
   ...:

In [4]: subroutine(1.0, None) # warmup
Out[4]: 7

In [5]: subroutine(1.0, None)
Out[5]: 1

In [6]: subroutine(1.0, ())
Out[6]: array(1)

In [7]: subroutine(1.0, (10,))
Out[7]: array([0, 0, 3, 3, 1, 1, 1, 2, 3, 1])

In [8]: subroutine(1.0, (10,10))
Out[8]:
array([[1, 3, 2, 0, 0, 1, 1, 1, 1, 0],
       [1, 0, 2, 0, 1, 1, 0, 1, 1, 0],
       [2, 1, 2, 1, 0, 2, 1, 1, 1, 0],
       [2, 2, 1, 3, 0, 3, 1, 2, 0, 1],
       [1, 0, 2, 0, 1, 0, 1, 0, 0, 0],
       [0, 0, 3, 1, 2, 0, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 3, 0, 0, 2],
       [0, 0, 1, 2, 2, 0, 2, 1, 1, 1],
       [0, 1, 0, 2, 0, 3, 0, 0, 0, 0],
       [0, 1, 3, 1, 0, 0, 1, 2, 1, 3]])
```

I tried all the requested styles: `None, (), (int,), (int, int)` above and they seemed to work fine.

**Comment by esc** [comment]
> I tried all the requested styles: `None, (), (int,), (int, int)` above and they seemed to work fine.

I ended up simply increasing the test coverage in the PR:

https://github.com/numba/numba/pull/9875

**Comment by esc** [comment]
I have updated the title to better reflect the ask here. I am currently assuming that the lack of tests as erroneously lead to the conclusion that multidimensional sampling was not implement, but it was. Thus I concluded that what is actually needed is increased test coverage for this distribution and have modified the related PR accordingly.

**Comment by esc** [comment]
I took another look and all the features requested by this PR were already implemented: https://github.com/numba/numba/pull/9875#issuecomment-2976734285

I will close the issue now.
